b2b-app/imports/ui/forms/survey-builder/recoil/atoms/parts-state.js

```diff
@@ -25,3 +31,182 @@ export const partsAtom = atom({
     },
   ],
 })
+
+const defaultPart = { name: '' }
+
+export const partAtom = atomFamily({
+  key: 'partAtom',
+  default: () => ({
+    prompt: '',
+    id: makeId(),
+    type: 'single',
+    answers: [makeListItem(defaultPart)],
+  }),
+  effects_UNSTABLE: (pid) => [
+    ({ setSelf }) => {
+      const part = dataCache.getPart(pid)
+      if (!part) return
+      setSelf(part)
+    },
+  ],
+})
+
+export const partAnswers = selectorFamily({
+  key: 'partAnswers',
+  get:
+    (pid) =>
+    ({ get }) => {
+      const part = get(partAtom(pid))
+      return part.answers
+    },
+  set:
+    (pid) =>
+    ({ get, set }, newValue) => {
+      const part = get(partAtom(pid))
+      const nextState = produce(part, (draft) => {
+        draft.answers = newValue
+      })
+
+      set(partAtom(pid), nextState)
+      // set(
+      //   partsAtom,
+      //   parts.map((part) => (part._id === pid ? nextState : part))
+      // )
+    },
+})
+
+export const partGridColumns = selectorFamily({
+  key: 'partGridColumns',
+  get:
+    (pid) =>
+    ({ get }) => {
+      const part = get(partAtom(pid))
+      return part.answers[0]?.columns
+    },
+  set:
+    (pid) =>
+    ({ get, set }, newValue) => {
+      const part = get(partAtom(pid))
+      const nextState = produce(part, (draft) => {
+        draft.answers[0].columns = newValue
+      })
+      set(partAtom(pid), nextState)
+    },
+})
+
+export const partGridRows = selectorFamily({
+  key: 'partGridRows',
+  get:
+    (pid) =>
+    ({ get }) => {
+      const part = get(partAtom(pid))
+      return part.answers[0]?.rows
+    },
+  set:
+    (pid) =>
+    ({ get, set }, newValue) => {
+      const part = get(partAtom(pid))
+      const nextState = produce(part, (draft) => {
+        draft.answers[0].rows = newValue
+      })
+      set(partAtom(pid), nextState)
+    },
+})
+
+export const editPartState = selectorFamily({
+  key: 'editPart',
+  get:
+    ({ pid, path }) =>
+    ({ get }) => {
+      // const parts = get(partsAtom)
+      const part = get(partAtom(pid))
+      // const part = findById(parts, pid)
+      return lget(part, path)
+    },
+  set:
+    ({ pid, path }) =>
+    ({ get, set }, newValue) => {
+      // const parts = get(partsAtom)
+      // const part = findById(parts, pid)
+      const part = get(partAtom(pid))
+      const nextState = produce(part, (draft) => {
+        lset(draft, path, newValue)
+      })
+      set(partAtom(pid), nextState)
+    },
+})
+
+//force rerender canvas when new question type is "section" because we need to re-categorise each questions to a specific section
+export const editPartsState = selectorFamily({
```

Comment:
Only update the whole list of questions when we want to re-render the canvas components. so, update the "partsAtom" not "partAtom".